### ISSUE TITLE
Hardcoded CUDA libraries in installed Legion CMake files, problems with CUDA forward-compatibility

### ISSUE THREAD

**Issue by rbberger** [StanfordLegion/legion#1299]
I’m working on the CI of FleCSI and ran into this issue while comparing builds for x86_64 and power9. We have power9 nodes with CUDA 11.0 drivers, and x86_64 nodes with CUDA 11.4 drivers. In both cases we’re using CUDA 11.7 with forward-compatibility. What that means is `libcuda.so` isn’t taken from `/usr/lib64/libcuda.so` at runtime but is injected through LD_LIBRARY_PATH. At link time the preference is to use the CUDA stubs libraries of the detected CUDA, to support building on nodes that don’t have CUDA drivers installed.

In the power9 case the node does have CUDA drivers installed, but since `/usr/lib64/libcuda.so` is a bit older, it doesn’t have all the symbols. So, you’ll get linking errors if you use that one instead of the CUDA 11.7 stubs or compat libraries.

```
lib64/librealm.so.1: undefined reference to `cuIpcOpenMemHandle_v2'
```

Now both Kokkos and Legion work by themselves. they build, install and link without issues. But the problems start when using Legion’s installed CMake files. **They inject hard-coded CUDA library paths into their CMake targets.**

In our CI, both Kokkos and Legion have been installed via Spack as dependencies. FleCSI then includes Legion’s CMake. Since the installed Legion CMake files have hardcoded the CUDA libraries in their targets, this leads to linking errors due to first finding `/usr/lib64/libcuda.so` instead of the CUDA 11.7 stubs or compatibility lib.

A typical way to fix this in your CMake project it to pass `CMAKE_LIBRARY_PATH=/path/to/stubs/or/compat` so CMake finds the stubs or compat libs before the `/usr/lib64` ones. However, this only works if all targets detect CUDA targets at configuration time.

Kokkos has the necessary logic to make this work. Their installed CMake files use `find(CUDAToolkit)` and detect if CUDA targets already exist and use the `CUDA::cuda_driver` and `CUDA::cudart` targets instead. Otherwise, they define them with absolute paths. Here is an excerpt of their installed CMake files:

**lib64/cmake/Kokkos/KokkosConfig.cmake (installed):**
```cmake
IF(NOT TARGET CUDA::cudart)
ADD_LIBRARY(CUDA::cudart INTERFACE IMPORTED)
SET_TARGET_PROPERTIES(CUDA::cudart PROPERTIES
INTERFACE_INCLUDE_DIRECTORIES "/path/to/cuda/include"
INTERFACE_LINK_LIBRARIES "/path/to/cuda/lib64/libcudart.so"
)
ENDIF()
IF(NOT TARGET CUDA::cuda_driver)
ADD_LIBRARY(CUDA::cuda_driver INTERFACE IMPORTED)
SET_TARGET_PROPERTIES(CUDA::cuda_driver PROPERTIES
INTERFACE_INCLUDE_DIRECTORIES "/path/to/cuda/include"
INTERFACE_LINK_LIBRARIES "/usr/lib64/libcuda.so"
)
```

Legion on the other hand has two places where it injects the hard-coded CUDA paths. One is the embedded GASNET installation, the other is Legion’s Runtime itself. It also uses the deprecated `find(CUDA)` and uses its own `FindCUDA` on top of it. This should become `find(CUDAToolkit)` to better take care of the corner cases. I’m also seeing CMake logic to remove `cudart` from `CUDA_LIBRARIES`. That should go away by only linking to `CUDA::cuda_driver` and not include `CUDA::cudart`.

**share/Legion/cmake/LegionTargets.cmake: (installed)**
```cmake
set_target_properties(Legion::LegionRuntime PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/path/to/cuda/include;${_IMPORT_PREFIX}/include;${_IMPORT_PREFIX}/include/mappers"
  INTERFACE_LINK_LIBRARIES "/path/to/cuda/lib64/libcudart.so;/usr/lib64/libcuda.so;Legion::RealmRuntime"
)
```

**share/Legion/cmake/LegionConfigCommon.cmake (installed):**
```cmake
if("${Legion_NETWORKS}" MATCHES ".*gasnet(1|ex).*")
  # define a GASNet::GASNet package that just points at:
  #  gasnet libs (embedded or external) for static builds
  #  external libs (e.g. mpi) needed in all cases
  cmake_minimum_required(VERSION 3.11)  # required for adding libs to imported target
  add_library(GASNet::GASNet INTERFACE IMPORTED)
  target_link_libraries(GASNet::GASNet INTERFACE /usr/lib64/libibverbs.so;/usr/lib64/libpmi2.so;/path/to/hwloc-2.7.1/lib/libhwloc.so;/usr/lib64/libcuda.so;/usr/lib64/libpthread.so;/usr/lib64/libm.so;/path/to/openmpi-4.1.2/lib/libmpi.so)
endif()
```

The way I solved the linking issues was by changing the Legion CMake to follow a similar strategy to Kokkos. Use `find(CUDAToolkit)` instead of the deprecated `find(CUDA)` and its targets `CUDA::cuda_driver` and `CUDA::cudart`. If they’re not defined, fall back to to providing absolute paths.

**Comment by lightsighter** [comment]
Someone who is more of expert on Legion's cmake build than me will need to comment on this issue, but I can say with some certainty that you're actually misunderstanding the motivation for this part:

> I’m also seeing CMake logic to remove cudart from CUDA_LIBRARIES. That should go away by only linking to CUDA::cuda_driver and not include CUDA::cudart.

We need to enforce a stronger invariant: we need to guarantee that nobody anywhere in the build is using libcudart.so because librealm.so exports its own version of the symbols in libcudart.so and we need everybody to link against that in place of libcudart.so (or things will break). If you know of a better way to enforce that invariant in cmake then I think we would be open to considering it, but it's not so simple as just making it so that liblegion.so and librealm.so don't link against libcudart.so.

**Comment by rbberger** [comment]
@lightsighter you're right. I didn't fully consider the cudart hijack parts and nor do I claim to fully understand them (the builds I've seen so far didn't activate that feature). Neither did I expect it to be that simple. 

When I created this issue I was expecting you have other constraints and requirements that I'm unaware of. I created it anyway to document and point out deficiencies and start a conversation.

Based on the CMake I'm seeing: if you want to enforce this stronger invariant, you will have to consider people not using `find(CUDA)` moving forward. Manipulating globals in the parent scope won't be enough.
